server/src/modules/auth/auth.service.ts

```diff
@@ -1,40 +1,174 @@
+    if (!user) {
+      return new ErrorResponse(`User with id: "${dto._id}" not found.`);
+    }
+
+    const userByEmail = await this.usersService.findOne({ email: dto.email });
+    if (userByEmail && !userByEmail._id.equals(dto._id)) {
+      return new ErrorResponse(
+        `User with email address: "${dto.email}" already exists.`,
+      );
+    }
+
+    if (user.password) {
+      if (user.email === dto.email) {
+        return new ErrorResponse(`Your account is already registered.`);
+      } else if (user.email) {
+        return new ErrorResponse(
+          `Your account is already registered with different email address.`,
+        );
+      }
+    }
+
+    dto.password = await hashPassword(dto.password);
+    Object.assign(user, dto);
+
+    await this.usersService.edit(user);
+
+    const getTokenInput: GetTokenInput = {
+      grantType: GrantType.AccessToken,
+      userName: `${user.userName}-${user._id}`,
+    };
+
+    return this.getToken(getTokenInput);
+  }
+
+  async login(dto: LoginUserInput): Promise<Token | ErrorResponse> {
+    const user = await this.usersService.findOne({ email: dto.email });
+    if (!user) {
+      return new ErrorResponse(
+        `User with email address: "${dto.email}" not found.`,
+      );
+    }
+
+    if (!user.password) {
+      return new ErrorResponse(
+        `User with email address: "${dto.email}" is not registered.`,
+      );
+    }
+
+    if (!(await verifyPassword(user.password, dto.password))) {
+      return new ErrorResponse(`Wrong email or password.`);
+    }
+
+    const getTokenInput: GetTokenInput = {
+      grantType: GrantType.AccessToken,
+      userName: `${user.userName}-${user._id}`,
+    };
+
+    return this.getToken(getTokenInput);
+  }
+
+  async loginByMagicLink(id: string): Promise<Token | ErrorResponse> {
+    const user = await this.usersService.findOne({ magicLinkId: id });
+    if (!user) {
+      return new ErrorResponse(`User for specified link does not exist.`);
+    }
+
+    const mlTime = Number(process.env.MAGIC_LINK_TIME);
+    const dateNow = new Date();
+    const createdDate = new Date(user.magicLinkCreatedAt);
+
+    createdDate.setSeconds(
+      createdDate.getSeconds() + mlTime ? mlTime : 300, // 5 minutes default
+    );
+
+    if (createdDate < dateNow) {
+      return new ErrorResponse(`Link has expired.`);
+    }
+
+    const getTokenInput: GetTokenInput = {
+      grantType: GrantType.AccessToken,
+      userName: `${user.userName}-${user._id}`,
+    };
+
+    return this.getToken(getTokenInput);
+  }
+
+  async getMagicLink(
+    magicLinkParam: MagicLinkInput,
+  ): Promise<string | ErrorResponse> {
+    const user = await this.usersService.findOne({
+      email: magicLinkParam.email,
+    });
+
+    if (!user || !user.password) {
+      return new ErrorResponse(
+        `User with email: "${magicLinkParam.email}" is not registered.`,
+      );
+    }
+
+    const linkId = uuid();
+
+    user.email = magicLinkParam.email;
```

Comment:
I don't get it... if you found the user by this email, why are you set this email again?